osutil/disks/gpt.go

```diff
@@ -0,0 +1,153 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package disks
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash/crc32"
+	"os"
+)
+
+type GPTLBA uint64
+type GPTGUID [16]byte
+
+// https://en.wikipedia.org/wiki/GUID_Partition_Table#Partition_table_header_(LBA_1)
+type GPTHeader struct {
+	Signature      [8]byte
+	Revision       uint32
+	HeaderSize     uint32
+	HeaderCRC      uint32
+	Reserved       uint32
+	CurrentLBA     GPTLBA
+	AlternateLBA   GPTLBA
+	FirstUsableLBA GPTLBA
+	LastUsableLBA  GPTLBA
+	DiskGUID       [16]byte
+	EntriesLBA     GPTLBA
+	NEntries       uint32
+	EntrySize      uint32
+	EntriesCRC     uint32
+}
+
+func verifyHeader(rawHeader []byte, header GPTHeader) error {
+	if !bytes.Equal(header.Signature[:], []byte("EFI PART")) {
+		return fmt.Errorf("GPT Header does not start with the magic string")
+	}
+
+	if header.Revision != 1<<16 {
+		return fmt.Errorf("GPT header revision is not 1.0")
+	}
+
+	if int(header.HeaderSize) < binary.Size(header) {
+		return fmt.Errorf("GPT header size is smaller than the minimum valid size")
+	}
+
+	if int(header.HeaderSize) > len(rawHeader) {
+		return fmt.Errorf("GPT header size is larger than the maximum supported size")
+	}
+
+	for i := 0; i < 4; i++ {
```

Comment:
this is clearing the Reserved field in the raw header correct? why do we need to clear it here? As per the spec it should always be zero, no? if it's not zero then the GPT isn't valid?

Reply:
It contains the crc. But the crc is calculated with the stored crc being 0.